### PR TITLE
Add multipart form-data transformations

### PR DESCRIPTION
**Pull request by tchambard**
The API remains the same, but the parser/formatter algorithm depends on multipart sub type.

Two sub types can now be managed : 
- mixed
- form-data

### CODE REVIEW

**Comment by tchambard** [comment]
@bjouhier I don't know why the pipeline failed...
Could you have a look please ?

**Comment by bjouhier** [comment]
@tchambard Thanks.

One of the unit tests timed out. Probably travis being a bit too busy. I just restarted the job and it succeeded. I will review so don't merge right away.

File: lib/transforms/multipart.ts [hunk 1]
```diff
@@ -3,17 +3,24 @@
...
+import { handshake } from 'f-promise';
 import * as generic from '../devices/generic';
 import * as binary from '../helpers/binary';
 import { Reader } from '../reader';
 import { Writer } from '../writer';
 
-function parseContentType(contentType?: string) {
+type MultipartSubType = 'mixed' | 'form-data';
+
+interface IMultipartContentType {
```
Comment (bjouhier) [inline lib/transforms/multipart.ts]: Can you rename to `MultipartContentType`. The `I` prefix for interfaces is unusual in TypeScript.

File: lib/transforms/multipart.ts [hunk 2]
```diff
@@ -25,70 +32,213 @@ function parseContentType(contentType?: string) {
...
+            const str = buf.toString('binary');
+
+            let endBoundaryIndex = str.indexOf(boundary) + boundary.length;
+            if (str.charAt(endBoundaryIndex) === '\r') {
+                endBoundaryIndex++;
+            }
+            if (str.charAt(endBoundaryIndex) === '\n') {
+                endBoundaryIndex++;
+            }
+            if (endBoundaryIndex < 0) throw new Error('boundary not found');
```
Comment (bjouhier) [inline lib/transforms/multipart.ts]: shouldn't this test be just after `let endBoundaryIndex = ...`?
Actually, before the `+ boundary.length` increment.

File: lib/transforms/multipart.ts [hunk 3]
```diff
@@ -25,70 +32,213 @@ function parseContentType(contentType?: string) {
...
+            }
+
+            binReader.unread(buf.length - beginOfData);
+
+            const read = () => {
+                const len = Math.max(boundary.length, 256);
+                const bbuf = binReader.readData(32 * len);
+                if (!bbuf || !bbuf.length) {
+                    hk.notify();
+                    return;
```
Comment (bjouhier) [inline lib/transforms/multipart.ts]:
```suggestion
                    return undefined;
```

File: test/unit/multipart-test.ts [hunk 4]
```diff
@@ -1,18 +1,17 @@
...
-import { run, wait } from 'f-promise';
 import { bufferReader, bufferWriter, multipartFormatter, multipartParser } from '../..';
 import { IncomingHttpHeaders } from 'http';
 setup();
 
-const { equal, ok, strictEqual, deepEqual } = assert;
+const { equal, ok, strictEqual } = assert;
 
-const boundary = '-- my boundary --';
+const boundary = '------------myBoundary';
```
Comment (bjouhier) [inline test/unit/multipart-test.ts]: Why did you have to change the boundary?

File: test/unit/multipart-test.ts [hunk 5]
```diff
@@ -79,9 +108,50 @@ describe(module.id, () => {
...
+        strictEqual(part.headers.b, 'VB2', 'header B');
+        strictEqual(part.headers['content-type'], 'text/plain', 'content-type');
+        strictEqual(part.headers['content-disposition'], 'form-data; name="c2";', 'content-disposition');
+        r = part.read();
+        strictEqual(r.toString('binary'), 'C2', 'body C2');
+        r = part.read();
+        strictEqual(r, undefined, 'end of part 2');
+
+        part = stream.read();
+        equal(part, undefined, 'read next part returns undefined');
```
Comment (bjouhier) [inline test/unit/multipart-test.ts]:
```suggestion
        assert.isUndefined(part);
```
Note: I omit the message argument when it paraphrases the test

File: lib/transforms/multipart.ts (hunk 1 above)
Comment (tchambard) [reply lib/transforms/multipart.ts]: done

File: lib/transforms/multipart.ts (hunk 3 above)
Comment (tchambard) [reply lib/transforms/multipart.ts]: done

File: test/unit/multipart-test.ts (hunk 4 above)
Comment (tchambard) [reply test/unit/multipart-test.ts]: The format seems to be more standard like that...

The double dash at the end is used normally to specify the end of the last content boundary

File: test/unit/multipart-test.ts (hunk 5 above)
Comment (tchambard) [reply test/unit/multipart-test.ts]: done

File: lib/transforms/multipart.ts (hunk 2 above)
Comment (tchambard) [reply lib/transforms/multipart.ts]: done

**Comment by bjouhier** [comment]
Hi Teddy. I merged and published. I bumped minor (new feature) and patch too because I had forgotten to merge my previous fix to master. Current version is now 2.2.2.
Thanks again.

**Comment by tchambard** [comment]
Perfect. Thank you very much Bruno.
One more question: do you think f-mongodb could be public or do you want to keep it private ?

**Comment by bjouhier** [comment]
Yes, I can publish it. You need it quickly?

**Comment by tchambard** [comment]
No, it's not an emergency, but if you do it at the occasion, I would probably try it :)

**Pull request merged**